File: tests/typescript/steps/address.steps.ts
import { Given, When, Then } from "@cucumber/cucumber";
import { expect } from "chai";
import { AccountAddress, Deserializer, Serializer } from "@aptos-labs/ts-sdk";
import type { AptosWorld } from "../support/world.js";
import {
  getAddressParsingVectors,
  getAddressConstants,
  getInvalidAddressInputs,
  hexToBytes,
  bytesToHex,
} from "../support/vectors.js";

// =============================================================================
// Given Steps - Address Input
// =============================================================================

// Primary step used in feature files
Given("a hex string {string}", function (this: AptosWorld, hex: string) {
  this.hexString = hex;
});

// Alternative phrasings for compatibility
Given(
  "a valid hex address string {string}",
  function (this: AptosWorld, hex: string) {
    this.hexString = hex;
  },
);

Given(
  "the address string {string}",
  function (this: AptosWorld, address: string) {
    this.hexString = address;
  },
);

Given("a short address {string}", function (this: AptosWorld, address: string) {
  this.hexString = address;
});

Given("a full 64-character hex address", function (this: AptosWorld) {
  this.hexString =
    "0x0000000000000000000000000000000000000000000000000000000000000001";
});

Given(
  "an invalid hex string {string}",
  function (this: AptosWorld, hex: string) {
    this.hexString = hex;
  },
);

Given("test vectors from addresses.json", function (this: AptosWorld) {
  const vectors = getAddressParsingVectors();
  this.testVectors.set("address_parsing", vectors);
});

// =============================================================================
// Given Steps - Address Creation from Value/Hex
// =============================================================================

Given(
  "an AccountAddress with value {int}",
  function (this: AptosWorld, value: number) {
    // Create address with specific byte value in last position
    const bytes = new Uint8Array(32);
    bytes[31] = value;
    this.address = AccountAddress.from(bytes);
  },
);

Given(
  "an AccountAddress from hex {string}",
  function (this: AptosWorld, hex: string) {
    this.address = AccountAddress.from(normalizeHexAddress(hex));
    // Store for comparison - initialize addresses array
    this.addresses = [this.address];
    this.testVectors.set("original_address", this.address);
  },
);

Given(
  "another AccountAddress from hex {string}",
  function (this: AptosWorld, hex: string) {
    const address2 = AccountAddress.from(normalizeHexAddress(hex));
    // Add second address for comparison
    if (!this.addresses || this.addresses.length === 0) {
      this.addresses = [this.address!];
    }
    this.addresses.push(address2);
  },
);

// =============================================================================
// Given Steps - Bytes
// =============================================================================

Given("32 random bytes", function (this: AptosWorld) {
  this.bytes = new Uint8Array(32);
  crypto.getRandomValues(this.bytes);
});

Given(
  "32 bytes with value {int} in the last byte",
  function (this: AptosWorld, value: number) {
    this.bytes = new Uint8Array(32);
    this.bytes[31] = value;
  },
);

// =============================================================================
// Given Steps - Address Constants
// =============================================================================

Given("the ZERO address constant", function (this: AptosWorld) {
  this.address = AccountAddress.ZERO;
});

Given("the zero address constant", function (this: AptosWorld) {
  this.address = AccountAddress.ZERO;
});

Given("the ONE address constant", function (this: AptosWorld) {
  this.address = AccountAddress.ONE;
});

Given("the framework address constant", function (this: AptosWorld) {
  this.address = AccountAddress.ONE;
});

Given("the THREE address constant", function (this: AptosWorld) {
  this.address = AccountAddress.THREE;
});

Given("the FOUR address constant", function (this: AptosWorld) {
  this.address = AccountAddress.FOUR;
});

// =============================================================================
// Given Steps - Address Pairs for Comparison
// =============================================================================

Given(
  "two addresses {string} and {string}",
  function (this: AptosWorld, addr1: string, addr2: string) {
    this.addresses = [AccountAddress.from(addr1), AccountAddress.from(addr2)];
  },
);

// =============================================================================
// When Steps - Parsing
// =============================================================================

// Helper to normalize short hex addresses for parsing
function normalizeHexAddress(hex: string): string {
  if (hex.startsWith("0x") || hex.startsWith("0X")) {
    const hexPart = hex.slice(2);
    if (hexPart.length < 64 && hexPart.length > 0) {
      return "0x" + hexPart.padStart(64, "0");
    }
    return hex;
  } else if (hex.length < 64 && hex.length > 0) {
    return "0x" + hex.padStart(64, "0");
  }
  return hex;
}

// Helper to get short string format (strip leading zeros)
// The TS SDK only provides short format for special addresses (0x1-0x4)
// So we implement our own for testing purposes
function toShortString(address: AccountAddress): string {
  const full = address.toStringLong();
  const hex = full.slice(2); // Remove 0x prefix
  const trimmed = hex.replace(/^0+/, ""); // Remove leading zeros
  return "0x" + (trimmed || "0"); // Ensure at least '0x0' for zero address
}

When("I parse the address", function (this: AptosWorld) {
  try {
    this.address = AccountAddress.from(normalizeHexAddress(this.hexString!));
    this.clearError();
  } catch (error) {
    this.setError(error as Error);
  }
});

When("I parse the address string", function (this: AptosWorld) {
  try {
    this.address = AccountAddress.from(normalizeHexAddress(this.hexString!));
    this.clearError();
  } catch (error) {
    this.setError(error as Error);
  }
});

When("I parse it as an AccountAddress", function (this: AptosWorld) {
  try {
    this.address = AccountAddress.from(normalizeHexAddress(this.hexString!));
    this.clearError();
  } catch (error) {
    this.setError(error as Error);
  }
});

When("I create an AccountAddress from the bytes", function (this: AptosWorld) {
  try {
    this.address = AccountAddress.from(this.bytes!);
    this.clearError();
  } catch (error) {
    this.setError(error as Error);
  }
});

When("I try to parse it as an address", function (this: AptosWorld) {
  try {
    this.address = AccountAddress.from(this.hexString!);
    this.clearError();
  } catch (error) {
    this.setError(error as Error);
  }
});

// =============================================================================
// When Steps - Formatting
// =============================================================================

When("I format the address as a string", function (this: AptosWorld) {
  this.hexString = this.address!.toString();
});

When("I format it as full hex", function (this: AptosWorld) {
  this.result = this.address!.toStringLong();
});

When("I format it as a full hex string", function (this: AptosWorld) {
  this.hexString = this.address!.toStringLong();
});

When("I format it as short string", function (this: AptosWorld) {
  this.result = toShortString(this.address!);
});

When("I format it as a short string", function (this: AptosWorld) {
  this.hexString = toShortString(this.address!);
});

// =============================================================================
// When Steps - Bytes/Serialization
// =============================================================================

When("I get the raw bytes", function (this: AptosWorld) {
  this.bytes = this.address!.toUint8Array();
});

When("I BCS serialize the address", function (this: AptosWorld) {
  // BCS serialization of an address is just the 32 bytes
  this.bytes = this.address!.toUint8Array();
});

When("I BCS deserialize as AccountAddress", function (this: AptosWorld) {
  try {
    const deserializer = new Deserializer(this.bytes!);
    this.address = AccountAddress.deserialize(deserializer);
    this.clearError();
  } catch (error) {
    this.setError(error as Error);
  }
});

When(
  "I BCS deserialize the result as AccountAddress",
  function (this: AptosWorld) {
    try {
      const deserializer = new Deserializer(this.bytes!);
      this.result = AccountAddress.deserialize(deserializer);
      this.clearError();
    } catch (error) {
      this.setError(error as Error);
    }
  },
);

// =============================================================================
// When Steps - Comparison
// =============================================================================

When("I compare them for equality", function (this: AptosWorld) {
  this.result = this.addresses[0].equals(this.addresses[1]);
});

// =============================================================================
// When Steps - Test Vectors
// =============================================================================

When("I run all parsing test vectors", function (this: AptosWorld) {
  const vectors = this.testVectors.get("address_parsing");
  const results: Array<{ name: string; passed: boolean; error?: string }> = [];

  for (const vector of vectors) {
    try {
      const address = AccountAddress.from(vector.input);
      const fullHex = address.toStringLong();
      const shortString = address.toString();

      const passed =
        fullHex.toLowerCase() === vector.expected.full_hex.toLowerCase() &&
        shortString.toLowerCase() ===
          vector.expected.short_string.toLowerCase();

      results.push({ name: vector.name, passed });
    } catch (error) {
      results.push({
        name: vector.name,
        passed: false,
        error: (error as Error).message,
      });
    }
  }

  this.result = results;
});

// =============================================================================
// Then Steps - Parsing Success/Failure
// =============================================================================

Then("the parsing should succeed", function (this: AptosWorld) {
  expect(this.error).to.be.undefined;
  // Check for either address or result (for TypeTags)
  expect(this.address ?? this.result).to.not.be.undefined;
});

Then("I should get a valid AccountAddress", function (this: AptosWorld) {
  expect(this.error).to.be.undefined;
  expect(this.address).to.not.be.undefined;
});

Then("the address should be valid", function (this: AptosWorld) {
  expect(this.error).to.be.undefined;
  expect(this.address).to.not.be.undefined;
});

Then("parsing should fail", function (this: AptosWorld) {
  expect(this.error).to.not.be.undefined;
});

Then("parsing should fail with an error", function (this: AptosWorld) {
  expect(this.error).to.not.be.undefined;
});

Then(
  "the parsing should fail with an invalid address error",
  function (this: AptosWorld) {
    expect(this.error).to.not.be.undefined;
  },
);

Then(
  "the parsing should fail with an invalid hex error",
  function (this: AptosWorld) {
    expect(this.error).to.not.be.undefined;
  },
);

Then(
  "the parsing should fail with an invalid length error",
  function (this: AptosWorld) {
    expect(this.error).to.not.be.undefined;
  },
);

Then(
  "it should fail with an InvalidAddress error",
  function (this: AptosWorld) {
    expect(this.error).to.not.be.undefined;
  },
);

// =============================================================================
// Then Steps - Address Byte Assertions
// =============================================================================

Then(
  "the address bytes should have length {int}",
  function (this: AptosWorld, length: number) {
    expect(this.address!.toUint8Array().length).to.equal(length);
  },
);

Then(
  "byte {int} should equal {int}",
  function (this: AptosWorld, index: number, value: number) {
    const bytes = this.address?.toUint8Array() ?? this.bytes;
    expect(bytes![index]).to.equal(value);
  },
);

Then(
  "bytes {int}-{int} should all be {int}",
  function (this: AptosWorld, start: number, end: number, value: number) {
    const bytes = this.address?.toUint8Array() ?? this.bytes;
    for (let i = start; i <= end; i++) {
      expect(bytes![i]).to.equal(value);
    }
  },
);

Then(
  "all 32 bytes should be {int}",
  function (this: AptosWorld, value: number) {
    const bytes = this.address!.toUint8Array();
    for (let i = 0; i < 32; i++) {
      expect(bytes[i]).to.equal(value);
    }
  },
);

Then("the byte length should be 32", function (this: AptosWorld) {
  expect(this.bytes!.length).to.equal(32);
});

Then("the bytes should be 32 bytes", function (this: AptosWorld) {
  expect(this.bytes!.length).to.equal(32);
});

// Removed duplicate - use regex version from hashing.steps.ts

Then("the first byte should be zero", function (this: AptosWorld) {
  expect(this.bytes![0]).to.equal(0);
});

Then(
  "the last byte should be {int}",
  function (this: AptosWorld, expected: number) {
    const bytes = this.address?.toUint8Array() ?? this.bytes;
    expect(bytes![31]).to.equal(expected);
  },
);

// =============================================================================
// Then Steps - String Format Assertions
// =============================================================================

Then(
  "the full hex should be {string}",
  function (this: AptosWorld, expected: string) {
    const actual = this.address!.toStringLong();
    expect(actual.toLowerCase()).to.equal(expected.toLowerCase());
  },
);

Then(
  "the full hex representation should be {string}",
  function (this: AptosWorld, expected: string) {
    const actual = this.address!.toStringLong();
    expect(actual.toLowerCase()).to.equal(expected.toLowerCase());
  },
);

Then(
  "the short string should be {string}",
  function (this: AptosWorld, expected: string) {
    const actual = toShortString(this.address!);
    expect(actual.toLowerCase()).to.equal(expected.toLowerCase());
  },
);

Then(
  "the address should equal {string}",
  function (this: AptosWorld, expected: string) {
    const actual = this.address!.toString();
    expect(actual.toLowerCase()).to.equal(expected.toLowerCase());
  },
);

Then(
  "the result should be {string}",
  function (this: AptosWorld, expected: string) {
    // Check testVectors.formattedString first, then this.result, then this.hexString
    const actual =
      this.testVectors.get("formattedString") ?? this.result ?? this.hexString;
    expect(String(actual).toLowerCase()).to.equal(expected.toLowerCase());
  },
);

// =============================================================================
// Then Steps - Address Equality
// =============================================================================

Then(
  "it should equal address {string}",
  function (this: AptosWorld, expected: string) {
    const expectedAddress = AccountAddress.from(expected);
    expect(this.address!.equals(expectedAddress)).to.be.true;
  },
);

Then("the two addresses should be equal", function (this: AptosWorld) {
  expect(this.addresses[0].equals(this.addresses[1])).to.be.true;
});

Then("the two addresses should not be equal", function (this: AptosWorld) {
  expect(this.addresses[0].equals(this.addresses[1])).to.be.false;
});

Then("they should be equal", function (this: AptosWorld) {
  // Handle both comparison result (boolean) and hash comparison (testVectors)
  if (this.testVectors.has("hash1") && this.testVectors.has("hash2")) {
    const hash1 = this.testVectors.get("hash1") as Uint8Array;
    const hash2 = this.testVectors.get("hash2") as Uint8Array;
    const hex1 = Array.from(hash1)
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");
    const hex2 = Array.from(hash2)
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");
    expect(hex1).to.equal(hex2);
  } else {
    expect(this.result).to.be.true;
  }
});

Then("they should not be equal", function (this: AptosWorld) {
  expect(this.result).to.be.false;
});

Then(
  "the result should equal the original address",
  function (this: AptosWorld) {
    const originalAddress = this.testVectors.get(
      "original_address",
    ) as AccountAddress;
    const resultAddress = this.result as AccountAddress;
    expect(resultAddress.equals(originalAddress)).to.be.true;
  },
);

// =============================================================================
// Then Steps - Test Vectors
// =============================================================================

Then("all test vectors should pass", function (this: AptosWorld) {
  const results = this.result as Array<{
    name: string;
    passed: boolean;
    error?: string;
  }>;
  const failures = results.filter((r) => !r.passed);

  if (failures.length > 0) {
    const failureMessages = failures
      .map((f) => `  - ${f.name}: ${f.error || "mismatch"}`)
      .join("\n");
    throw new Error(
      `${failures.length} test vectors failed:\n${failureMessages}`,
    );
  }
});

File: tests/python/steps/address_steps.py
"""
Step definitions for address.feature
Tests AccountAddress parsing, formatting, and BCS serialization.
"""

from support.vectors import (
    get_address_parsing_vectors,
    bytes_to_hex,
)
from aptos_sdk.bcs import Serializer, Deserializer
from aptos_sdk.account_address import AccountAddress
from behave import given, when, then
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))


# =============================================================================
# Helper Functions
# =============================================================================


def to_full_hex(address: AccountAddress) -> str:
    """Format address as full 64-character hex string."""
    return "0x" + address.address.hex()


def to_short_string(address: AccountAddress) -> str:
    """Format address as short string (strip leading zeros)."""
    full_hex = address.address.hex()
    trimmed = full_hex.lstrip("0") or "0"
    return "0x" + trimmed


def parse_address(hex_string: str) -> AccountAddress:
    """Parse an address using relaxed parsing."""
    return AccountAddress.from_str_relaxed(hex_string)


# =============================================================================
# Given Steps - Address Input
# =============================================================================


@given('a hex string "{hex_string}"')
def step_given_hex_string(context, hex_string):
    context.world.hex_string = hex_string


@given('a hex string ""')
def step_given_empty_hex_string(context):
    context.world.hex_string = ""


@given('a valid hex address string "{hex_string}"')
def step_given_valid_hex_address(context, hex_string):
    context.world.hex_string = hex_string


@given('the address string "{address}"')
def step_given_address_string(context, address):
    context.world.hex_string = address


@given('a short address "{address}"')
def step_given_short_address(context, address):
    context.world.hex_string = address


@given("a full 64-character hex address")
def step_given_full_hex_address(context):
    context.world.hex_string = (
        "0x0000000000000000000000000000000000000000000000000000000000000001"
    )


# Note: "an invalid hex string" is defined in account_steps.py


@given("test vectors from addresses.json")
def step_given_address_test_vectors(context):
    vectors = get_address_parsing_vectors()
    context.world.test_vectors["address_parsing"] = vectors


# =============================================================================
# Given Steps - Address Creation
# =============================================================================


@given("an AccountAddress with value {value:d}")
def step_given_address_with_value(context, value):
    # Create address with specific byte value in last position
    address_bytes = bytearray(32)
    address_bytes[31] = value
    full_hex = "0x" + bytes(address_bytes).hex()
    context.world.address = AccountAddress.from_str_relaxed(full_hex)


@given('an AccountAddress from hex "{hex_string}"')
def step_given_address_from_hex(context, hex_string):
    context.world.address = parse_address(hex_string)
    context.world.addresses = [context.world.address]
    context.world.test_vectors["original_address"] = context.world.address


@given('another AccountAddress from hex "{hex_string}"')
def step_given_another_address_from_hex(context, hex_string):
    address2 = parse_address(hex_string)
    if not context.world.addresses:
        context.world.addresses = [context.world.address]
    context.world.addresses.append(address2)


# =============================================================================
# Given Steps - Bytes
# =============================================================================


@given("32 random bytes")
def step_given_random_bytes(context):
    import os

    context.world.bytes_value = os.urandom(32)


@given("32 bytes with value {value:d} in the last byte")
def step_given_bytes_with_value(context, value):
    data = bytearray(32)
    data[31] = value
    context.world.bytes_value = bytes(data)


# =============================================================================
# Given Steps - Address Constants
# =============================================================================


@given("the ZERO address constant")
def step_given_zero_address(context):
    context.world.address = AccountAddress.from_str(
        "0x0000000000000000000000000000000000000000000000000000000000000000"
    )


@given("the zero address constant")
def step_given_zero_address_alt(context):
    context.world.address = AccountAddress.from_str(
        "0x0000000000000000000000000000000000000000000000000000000000000000"
    )


@given("the ONE address constant")
def step_given_one_address(context):
    context.world.address = AccountAddress.from_str(
        "0x0000000000000000000000000000000000000000000000000000000000000001"
    )


@given("the framework address constant")
def step_given_framework_address(context):
    context.world.address = AccountAddress.from_str(
        "0x0000000000000000000000000000000000000000000000000000000000000001"
    )


@given("the THREE address constant")
def step_given_three_address(context):
    context.world.address = AccountAddress.from_str(
        "0x0000000000000000000000000000000000000000000000000000000000000003"
    )


@given("the FOUR address constant")
def step_given_four_address(context):
    context.world.address = AccountAddress.from_str(
        "0x0000000000000000000000000000000000000000000000000000000000000004"
    )


# =============================================================================
# Given Steps - Address Pairs
# =============================================================================


@given('two addresses "{addr1}" and "{addr2}"')
def step_given_two_addresses(context, addr1, addr2):
    context.world.addresses = [
        parse_address(addr1),
        parse_address(addr2),
    ]


# =============================================================================
# When Steps - Parsing
# =============================================================================


@when("I parse the address")
def step_parse_address(context):
    try:
        context.world.address = AccountAddress.from_str_relaxed(
            context.world.hex_string
        )
        context.world.clear_error()
    except Exception as e:
        context.world.set_error(e)


@when("I parse the address string")
def step_parse_address_string(context):
    step_parse_address(context)


@when("I parse it as an AccountAddress")
def step_parse_as_account_address(context):
    step_parse_address(context)


@when("I create an AccountAddress from the bytes")
def step_create_address_from_bytes(context):
    try:
        hex_str = bytes_to_hex(context.world.bytes_value)
        context.world.address = AccountAddress.from_str(hex_str)
        context.world.clear_error()
    except Exception as e:
        context.world.set_error(e)


@when("I try to parse it as an address")
def step_try_parse_address(context):
    try:
        context.world.address = AccountAddress.from_str_relaxed(
            context.world.hex_string
        )
        context.world.clear_error()
    except Exception as e:
        context.world.set_error(e)


# =============================================================================
# When Steps - Formatting
# =============================================================================


@when("I format the address as a string")
def step_format_address_as_string(context):
    context.world.hex_string = str(context.world.address)


@when("I format it as full hex")
def step_format_as_full_hex(context):
    context.world.result = to_full_hex(context.world.address)


@when("I format it as a full hex string")
def step_format_as_full_hex_string(context):
    context.world.hex_string = str(context.world.address)


@when("I format it as short string")
def step_format_as_short_string(context):
    context.world.result = to_short_string(context.world.address)


@when("I format it as a short string")
def step_format_as_short_string_alt(context):
    context.world.hex_string = to_short_string(context.world.address)


# =============================================================================
# When Steps - Bytes/Serialization
# =============================================================================


@when("I get the raw bytes")
def step_get_raw_bytes(context):
    # AccountAddress stores bytes internally
    context.world.bytes_value = bytes(context.world.address.address)


@when("I BCS serialize the address")
def step_bcs_serialize_address(context):
    serializer = Serializer()
    serializer.struct(context.world.address)
    context.world.bytes_value = serializer.output()


@when("I BCS deserialize as AccountAddress")
def step_bcs_deserialize_address(context):
    try:
        deserializer = Deserializer(context.world.bytes_value)
        context.world.address = AccountAddress.deserialize(deserializer)
        context.world.clear_error()
    except Exception as e:
        context.world.set_error(e)


@when("I BCS deserialize the result as AccountAddress")
def step_bcs_deserialize_result_as_address(context):
    try:
        deserializer = Deserializer(context.world.bytes_value)
        context.world.result = AccountAddress.deserialize(deserializer)
        context.world.clear_error()
    except Exception as e:
        context.world.set_error(e)


# =============================================================================
# When Steps - Comparison
# =============================================================================


@when("I compare them for equality")
def step_compare_addresses(context):
    addr1 = context.world.addresses[0]
    addr2 = context.world.addresses[1]
    context.world.result = str(addr1) == str(addr2)


# =============================================================================
# When Steps - Test Vectors
# =============================================================================


@when("I run all parsing test vectors")
def step_run_parsing_test_vectors(context):
    vectors = context.world.test_vectors.get("address_parsing", [])
    results = []

    for vector in vectors:
        try:
            address = AccountAddress.from_str(vector["input"])
            full_hex = str(address)
            short_str = to_short_string(address)

            passed = (
                full_hex.lower() == vector["expected"]["full_hex"].lower()
                and short_str.lower() == vector["expected"]["short_string"].lower()
            )
            results.append({"name": vector["name"], "passed": passed})
        except Exception as e:
            results.append({"name": vector["name"], "passed": False, "error": str(e)})

    context.world.result = results


# =============================================================================
# Then Steps - Parsing Success/Failure
# =============================================================================


@then("the parsing should succeed")
def step_parsing_should_succeed(context):
    assert context.world.error is None, f"Expected no error, got: {context.world.error}"
    # Check various things that might be set after parsing
    has_something = (
        context.world.address is not None
        or context.world.result is not None
        or getattr(context.world, "type_tag", None) is not None
        or getattr(context.world, "module_address", None) is not None
        or getattr(context.world, "hash_value", None) is not None
    )
    assert (
        has_something
    ), "Expected address, result, type_tag, module_address, or hash_value to be set"


@then("I should get a valid AccountAddress")
def step_should_get_valid_address(context):
    assert context.world.error is None
    assert context.world.address is not None


@then("the address should be valid")
def step_address_should_be_valid(context):
    assert context.world.error is None
    assert context.world.address is not None


@then("parsing should fail")
def step_parsing_should_fail(context):
    assert context.world.error is not None, "Expected an error but got none"


@then("parsing should fail with an error")
def step_parsing_should_fail_with_error(context):
    assert context.world.error is not None


@then("the parsing should fail with an invalid address error")
def step_parsing_should_fail_invalid_address(context):
    assert context.world.error is not None


@then("the parsing should fail with an invalid hex error")
def step_parsing_should_fail_invalid_hex(context):
    assert context.world.error is not None


@then("the parsing should fail with an invalid length error")
def step_parsing_should_fail_invalid_length(context):
    assert context.world.error is not None


@then("it should fail with an InvalidAddress error")
def step_should_fail_invalid_address(context):
    assert context.world.error is not None


# =============================================================================
# Then Steps - Address Byte Assertions
# =============================================================================


@then("the address bytes should have length {length:d}")
def step_address_bytes_length(context, length):
    addr_bytes = bytes(context.world.address.address)
    assert len(addr_bytes) == length


@then("byte {index:d} should equal {value:d}")
def step_byte_should_equal(context, index, value):
    if context.world.address:
        data = bytes(context.world.address.address)
    else:
        data = context.world.bytes_value
    assert data[index] == value


@then("bytes {start:d}-{end:d} should all be {value:d}")
def step_bytes_range_should_equal(context, start, end, value):
    if context.world.address:
        data = bytes(context.world.address.address)
    else:
        data = context.world.bytes_value
    for i in range(start, end + 1):
        assert data[i] == value, f"Byte {i} is {data[i]}, expected {value}"


@then("all 32 bytes should be {value:d}")
def step_all_bytes_should_equal(context, value):
    data = bytes(context.world.address.address)
    for i in range(32):
        assert data[i] == value


@then("the byte length should be 32")
def step_byte_length_should_be_32(context):
    assert len(context.world.bytes_value) == 32


@then("the bytes should be 32 bytes")
def step_bytes_should_be_32(context):
    assert len(context.world.bytes_value) == 32


@then("the result should be 32 bytes")
def step_result_should_be_32_bytes(context):
    assert len(context.world.bytes_value) == 32


@then("the first byte should be zero")
def step_first_byte_zero(context):
    assert context.world.bytes_value[0] == 0


@then("the last byte should be {expected:d}")
def step_last_byte_should_be(context, expected):
    if context.world.address:
        data = bytes(context.world.address.address)
    else:
        data = context.world.bytes_value
    assert data[31] == expected


# =============================================================================
# Then Steps - String Format Assertions
# =============================================================================


@then('the full hex should be "{expected}"')
def step_full_hex_should_be(context, expected):
    actual = to_full_hex(context.world.address)
    assert actual.lower() == expected.lower()


@then('the full hex representation should be "{expected}"')
def step_full_hex_representation_should_be(context, expected):
    actual = to_full_hex(context.world.address)
    assert actual.lower() == expected.lower()


@then('the short string should be "{expected}"')
def step_short_string_should_be(context, expected):
    actual = to_short_string(context.world.address)
    assert actual.lower() == expected.lower()


@then('the address should equal "{expected}"')
def step_address_should_equal(context, expected):
    actual = str(context.world.address)
    assert actual.lower() == expected.lower()


@then('the result should be "{expected}"')
def step_result_should_be(context, expected):
    actual = (
        context.world.test_vectors.get("formattedString")
        or context.world.result
        or context.world.hex_string
    )
    assert str(actual).lower() == expected.lower()


# =============================================================================
# Then Steps - Address Equality
# =============================================================================


@then('it should equal address "{expected}"')
def step_should_equal_address(context, expected):
    expected_addr = parse_address(expected)
    assert to_full_hex(context.world.address) == to_full_hex(expected_addr)


@then("the two addresses should be equal")
def step_two_addresses_equal(context):
    assert to_full_hex(context.world.addresses[0]) == to_full_hex(
        context.world.addresses[1]
    )


@then("the two addresses should not be equal")
def step_two_addresses_not_equal(context):
    assert to_full_hex(context.world.addresses[0]) != to_full_hex(
        context.world.addresses[1]
    )


@then("they should be equal")
def step_they_should_be_equal(context):
    if "hash1" in context.world.test_vectors and "hash2" in context.world.test_vectors:
        hash1 = context.world.test_vectors["hash1"]
        hash2 = context.world.test_vectors["hash2"]
        assert hash1.hex() == hash2.hex()
    else:
        assert context.world.result is True


@then("they should not be equal")
def step_they_should_not_be_equal(context):
    assert context.world.result is False


@then("the result should equal the original address")
def step_result_equals_original(context):
    original = context.world.test_vectors.get("original_address")
    result = context.world.result
    assert str(result) == str(original)


# =============================================================================
# Then Steps - Test Vectors
# =============================================================================


@then("all test vectors should pass")
def step_all_vectors_pass(context):
    results = context.world.result
    failures = [r for r in results if not r["passed"]]

    if failures:
        failure_msgs = [
            f"  - {f['name']}: {f.get('error', 'mismatch')}" for f in failures
        ]
        raise AssertionError(
            f"{len(failures)} test vectors failed:\n" + "\n".join(failure_msgs)
        )
